# Multivariate EDA

In this notebook I will perform multivariate analysis to see if there are more hidden relationships. 


Questions that I can think about right now and will be answered in this notebook:
1. Does vtat_missing and calcenation rates behave differently across vehicle types, locations, or time?
2. Does pickup × drop add signal beyond individual locations?
3. Confirm that avg_vtat missingness is not confounded
4. Correlation matrix & multicollinearity among engineered features
5. Non-linear feature relevance ranking
6. Deterministic vs probabilistic subset separation
7. Model assumption validation

## Setup & Data Loading

In [1]:
import sys, os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
from scipy.stats import chi2_contingency, fisher_exact
from sklearn.feature_selection import mutual_info_classif
from statsmodels.stats.outliers_influence import variance_inflation_factor

sys.path.insert(0, os.path.abspath("../src"))
from eda_utils.bivar_utils import cramers_v

pd.set_option("display.max_columns", 40)
plt.rcParams.update({"figure.dpi": 100, "figure.figsize": (12, 6)})



In [2]:
df = pd.read_parquet("../data/silver/data_to_multivar.parquet", engine='pyarrow')
df = df.sort_values(["date", "time"]).reset_index(drop=True)
df.head()

,date,time,vehicle_type,pickup_location,drop_location,avg_vtat,is_cancelled,weekday,month,day_of_month,week_of_year,quarter,time_parsed,hour,is_night,is_business,is_rush,vtat_zone,is_instant_arrival,is_timeout,is_long_wait,vtat_missing
0,2024-01-01,00:19:34,Bike,Udyog Vihar,Ambience Mall,10.8,0,0,1,1,1,1,1900-01-01 00:19:34,0,1,0,0,normal,0,0,0,0
1,2024-01-01,01:35:18,Go Mini,Basai Dhankot,Madipur,8.5,0,0,1,1,1,1,1900-01-01 01:35:18,1,1,0,0,normal,0,0,0,0
2,2024-01-01,01:37:50,Go Sedan,Tughlakabad,Greater Kailash,7.4,1,0,1,1,1,1,1900-01-01 01:37:50,1,1,0,0,normal,0,0,0,0
3,2024-01-01,01:48:03,Auto,Palam Vihar,Kherki Daula Toll,5.6,1,0,1,1,1,1,1900-01-01 01:48:03,1,1,0,0,normal,0,0,0,0
4,2024-01-01,01:49:56,Go Sedan,Narsinghpur,Pulbangash,6.2,1,0,1,1,1,1,1900-01-01 01:49:56,1,1,0,0,normal,0,0,0,0


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 22 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   date                150000 non-null  datetime64[ns]
 1   time                150000 non-null  object        
 2   vehicle_type        150000 non-null  category      
 3   pickup_location     150000 non-null  category      
 4   drop_location       150000 non-null  category      
 5   avg_vtat            139500 non-null  float32       
 6   is_cancelled        150000 non-null  int64         
 7   weekday             150000 non-null  int32         
 8   month               150000 non-null  int32         
 9   day_of_month        150000 non-null  int32         
 10  week_of_year        150000 non-null  int64         
 11  quarter             150000 non-null  int32         
 12  time_parsed         150000 non-null  datetime64[ns]
 13  hour                150000 no

## Multivariate Missingness Confounding Check

The bivariate analysis notebook showed that avg_vtat missingness is spread roughly equally across all independent features and I left it here to check whether combinations of these independent + dependent features can predict missingness. For that I decided to go with a log regression and I will compare it with the baseline! 

Yhe question here would be: can the combination of all non-VTAT features predict whether avg_vtat (imbalanced) is missing any better than is_cancelled alone? For that, AUC (robust to class imbalance) measures the model's overall ranking ability across all possible decision thresholds.

In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.preprocessing import LabelEncoder

non_vtat_cols = [
    "weekday", "month", "day_of_month", "week_of_year", "quarter",
    "hour", "is_night", "is_business", "is_rush",
    "vehicle_type", "pickup_location", "drop_location",
    "is_cancelled",
]

X = df[non_vtat_cols].copy()
for c in X.select_dtypes(include="category").columns:
    X[c] = LabelEncoder().fit_transform(X[c])
y = df["vtat_missing"]

lr_full = LogisticRegression(max_iter=2000, solver="lbfgs")
lr_full.fit(X, y)
y_prob_full = lr_full.predict_proba(X)[:, 1]
auc_full = roc_auc_score(y, y_prob_full)

X_base = df[["is_cancelled"]]
lr_base = LogisticRegression(max_iter=2000, solver="lbfgs")
lr_base.fit(X_base, y)
y_prob_base = lr_base.predict_proba(X_base)[:, 1]
auc_base = roc_auc_score(y, y_prob_base)

print(f"Baseline model (is_cancelled only)  AUC = {auc_base:.4f}")
print(f"Full model (all non-VTAT features)  AUC = {auc_full:.4f}")
print(f"AUC lift from adding all features   ΔAU = {auc_full - auc_base:.4f}")

coef_df = (
    pd.DataFrame({"feature": non_vtat_cols, "coef": lr_full.coef_[0]})
    .assign(abs_coef=lambda d: d["coef"].abs())
    .sort_values("abs_coef", ascending=False)
    .reset_index(drop=True)
)
print("\nLogistic regression coefficients (full model):")
coef_df

Baseline model (is_cancelled only)  AUC = 0.8656
Full model (all non-VTAT features)  AUC = 0.8675
AUC lift from adding all features   ΔAU = 0.0019

Logistic regression coefficients (full model):


,feature,coef,abs_coef
0,is_cancelled,8.243286,8.243286
1,is_night,-0.048240,0.048240
2,quarter,-0.031554,0.031554
3,is_business,-0.030319,0.030319
4,is_rush,-0.023010,0.023010
5,month,-0.004673,0.004673
6,weekday,0.004632,0.004632
7,week_of_year,0.003513,0.003513
8,hour,-0.003383,0.003383
9,vehicle_type,0.001909,0.001909


The full model and baseline model achieve almost the same result, so no combination of features predicts missingness beyond is_cancelled alone. The missingness mechanism is MNAR driven by the target with no hidden confounding from feature interactions.

That said, vtat_missing is not simply is_cancelled in disguise. Many cancelled rides do have a VTAT value recorded (the vehicle arrived, the rider still cancelled), and some non-cancelled rides have VTAT missing for unknown reasons, likely logging failures or system issues? 

Imputation of avg_vtat or using vtat_missing as a feature won't introduce bias from confounding with other features.

Since the dataset contains different original features and their engineered derivated, I'm going to group them in blocks and check whixh has the highest effect size over the target

In [5]:
feature_block = {
    "temporal_date": ["weekday", "month", "day_of_month", "week_of_year", "quarter"],
    "temporal_time": ["hour", "is_rush", "is_night", "is_business"],
    "pickup_location": ["pickup_location"],
    "drop_location": ["drop_location"],
    "vehicle": ["vehicle_type"],
    "vtat": ["avg_vtat", "vtat_zone", "is_instant_arrival" ,"is_timeout", "is_long_wait", "vtat_missing"], 
    "is_cancelled": ["is_cancelled"]
}

rows = []
for block, cols in feature_block.items():
    for col in cols:
        if col == "is_cancelled":
            continue
        rates = df.groupby(col)['is_cancelled'].mean()
        rows.append({
            "block": block,
            "feature": col,
            "rate_swing": rates.max() - rates.min(),
        })

rate_swing_df = pd.DataFrame(rows).sort_values(["block", "rate_swing"], ascending=False).reset_index(drop=True)
rate_swing_df


/tmp/ipykernel_28075/2652892690.py:16: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  rates = df.groupby(col)['is_cancelled'].mean()


,block,feature,rate_swing
0,vtat,avg_vtat,1.000000
1,vtat,vtat_zone,1.000000
2,vtat,vtat_missing,0.731183
3,vtat,is_timeout,0.696346
4,vtat,is_instant_arrival,0.337299
5,vtat,is_long_wait,0.135741
6,vehicle,vehicle_type,0.009504
7,temporal_time,hour,0.032106
8,temporal_time,is_night,0.004227
9,temporal_time,is_rush,0.003141


## Family Redundancy Check

Before doing any further multivariate work I need to quantify how much overlap there is and whether keeping all three adds signal or just noise.

temporal_tim: is_night, is_rush, is_business are deterministic binary flags derived from hour. Redundant by construction. I will pick hour as representative.

temporal_date: same situation as with temporal_time variables. All of them come from date, and it was proven in the previous notebook that is carried no signal. 

vehicle, pickup_location, drop_location are all single variable each. Nothing to check.

vtat family will be checked:

#### avg_vtat Family

In [6]:
complete = df.dropna(subset=["avg_vtat"]).copy()

groups = [g["avg_vtat"].values for _, g in complete.groupby("vtat_zone", observed=True)]
ss_bw = sum(len(g) * (g.mean() - complete["avg_vtat"].mean()) ** 2 for g in groups)
ss_tot = ((complete["avg_vtat"] - complete["avg_vtat"].mean()) ** 2).sum()
eta_sq_1 = ss_bw / ss_tot

zone_codes = complete["vtat_zone"].cat.codes.values.reshape(-1, 1).astype(float)
mi_1 = mutual_info_classif(
    complete[["avg_vtat"]].values,
    zone_codes.ravel().astype(int),
    discrete_features=[False], random_state=42, n_neighbors=5,
)[0]

print("avg_vtat - vtat_zone")
print(f"  eta-sq = {eta_sq_1:.4f}")
print(f"  MI     = {mi_1:.4f}")


cv_3 = cramers_v(df["vtat_zone"].astype(str), df["vtat_missing"].astype(str))

print("\nvtat_zone - vtat_missing")
print(f"  Cramer's V = {cv_3:.4f}")





avg_vtat - vtat_zone
  eta-sq = 0.8239
  MI     = 1.1667

vtat_zone - vtat_missing
  Cramer's V = 1.0000


vtat_zone is highly redundant (compare with avg_vtat in rate_swing_df above) so I will drop it and keep avg_vtat because it has smaller granularity, avg_vtat and vtat_missing are structurally complementary so I will keep both for linear models and (still to decide) eliminate vtat_missing while imputing sentinel value (-1) for trees



In [7]:
vars = df[["avg_vtat", "vehicle_type", "pickup_location", "drop_location", "vtat_missing", "is_cancelled"]].copy()
vars

,avg_vtat,vehicle_type,pickup_location,drop_location,vtat_missing,is_cancelled
0,10.8,Bike,Udyog Vihar,Ambience Mall,0,0
1,8.5,Go Mini,Basai Dhankot,Madipur,0,0
2,7.4,Go Sedan,Tughlakabad,Greater Kailash,0,1
3,5.6,Auto,Palam Vihar,Kherki Daula Toll,0,1
4,6.2,Go Sedan,Narsinghpur,Pulbangash,0,1
...,...,...,...,...,...,...
149995,6.9,Bike,DLF Phase 3,Okhla,0,0
149996,7.9,eBike,Saket,Noida Sector 62,0,0
149997,9.9,Go Mini,GTB Nagar,Anand Vihar ISBT,0,0
149998,2.6,Uber XL,Ashram,Vasant Kunj,0,0


## Interaction Effects: pickup × drop

Both locations had meaningful individual rate swings in the bivariate analysis. The question here is whether specific routes carry signal 

In [8]:
route = vars.copy()
route["route"] = route["pickup_location"].astype(str) + "_" + route["drop_location"].astype(str)

target = route["is_cancelled"].astype(str)

v_pickup = cramers_v(route["pickup_location"].astype(str), target)
v_drop   = cramers_v(route["drop_location"].astype(str), target)
v_route  = cramers_v(route["route"], target)

print(f"Cramer's V with target:")
print(f"  pickup_location: {v_pickup:.4f}")
print(f"  drop_location:   {v_drop:.4f}")
print(f"  route (pick_drop): {v_route:.4f}")


Cramer's V with target:
  pickup_location: 0.0369
  drop_location:   0.0367
  route (pick_drop): 0.4485


The route combination is a far stronger predictor than either location alone, 
both individual locations have near-zero association with the target while the combined route is substantial. However, Cramer's V is inflated by the high cardinality of route still the jump is too large to explain away, the specific origin-destination pair matters

## VIF

Will multicollinearity destabilize my linear model?


In [9]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif_df = vars.drop(columns=["is_cancelled"]).copy()

# Build route and frequency-encode it
vif_df["route"] = vif_df["pickup_location"].astype(str) + "_" + vif_df["drop_location"].astype(str)
route_freq = vif_df["route"].value_counts()
vif_df["route_freq"] = vif_df["route"].map(route_freq).astype(float)

# Frequency-encode individual locations
pickup_freq = vif_df["pickup_location"].value_counts()
drop_freq = vif_df["drop_location"].value_counts()
vif_df["pickup_freq"] = vif_df["pickup_location"].map(pickup_freq).astype(float)
vif_df["drop_freq"] = vif_df["drop_location"].map(drop_freq).astype(float)
vif_df = vif_df.drop(columns=["pickup_location", "drop_location", "route"])

# One-hot encode vehicle_type
vif_df = pd.get_dummies(vif_df, columns=["vehicle_type"], drop_first=True, dtype=float)

# Fill avg_vtat NaN with median
vif_df["avg_vtat"] = vif_df["avg_vtat"].fillna(vif_df["avg_vtat"].median())

vif_df.insert(0, "const", 1.0)

vif_results = pd.DataFrame({
    "feature": vif_df.columns[1:],
    "VIF": [variance_inflation_factor(vif_df.values, i) for i in range(1, vif_df.shape[1])]
})

print(vif_results.to_string(index=False))


                   feature      VIF
                  avg_vtat 1.000213
              vtat_missing 1.000234
                route_freq 1.012971
               pickup_freq 1.006771
                 drop_freq 1.006268
         vehicle_type_Bike 1.361382
      vehicle_type_Go Mini 1.439641
     vehicle_type_Go Sedan 1.413185
vehicle_type_Premier Sedan 1.304866
      vehicle_type_Uber XL 1.085748
        vehicle_type_eBike 1.191904


All VIFs close to 1, including route. No multicollinearity among any of the selected features. 
Route frequency doesn't inflate even though it's constructed from pickup and drop, 
because a route's popularity is not a simple sum of its endpoints' popularities. 
Safe to use all features together in any model family

## Mutual Information: Non-linear Dependency Check

VIF only captures linear redundancy. MI detects any dependency (linear or non-linear) between feature pairs. High MI between two features means one helps predict the other, even if VIF was low

In [10]:
from sklearn.feature_selection import mutual_info_regression, mutual_info_classif
import itertools

mi_df = vars.drop(columns=["is_cancelled"]).copy()

# Fill NaN in avg_vtat for MI computation
mi_df["avg_vtat"] = mi_df["avg_vtat"].fillna(mi_df["avg_vtat"].median())

# Encode categoricals as integers
mi_df["vehicle_type_enc"] = mi_df["vehicle_type"].cat.codes.astype(float)
mi_df["pickup_enc"] = mi_df["pickup_location"].cat.codes.astype(float)
mi_df["drop_enc"] = mi_df["drop_location"].cat.codes.astype(float)
mi_df["route"] = mi_df["pickup_location"].astype(str) + "_" + mi_df["drop_location"].astype(str)
mi_df["route_enc"] = pd.Categorical(mi_df["route"]).codes.astype(float)
mi_df = mi_df.drop(columns=["vehicle_type", "pickup_location", "drop_location", "route"])

features = ["avg_vtat", "vtat_missing", "vehicle_type_enc", "pickup_enc", "drop_enc", "route_enc"]
feature_labels = ["avg_vtat", "vtat_missing", "vehicle_type", "pickup", "drop", "route"]

mi_matrix = pd.DataFrame(index=feature_labels, columns=feature_labels, dtype=float)

for i, feat_i in enumerate(features):
    for j, feat_j in enumerate(features):
        if i == j:
            mi_matrix.iloc[i, j] = np.nan
        elif i < j:
            # Compute MI of feat_i given feat_j (treat feat_j as target)
            discrete = feat_j in ["vtat_missing", "vehicle_type_enc", "pickup_enc", "drop_enc", "route_enc"]
            if discrete:
                mi = mutual_info_classif(
                    mi_df[[feat_i]].values, 
                    mi_df[feat_j].astype(int).values,
                    discrete_features=[feat_i in ["vtat_missing", "vehicle_type_enc", "pickup_enc", "drop_enc", "route_enc"]],
                    random_state=42, n_neighbors=5
                )[0]
            else:
                mi = mutual_info_regression(
                    mi_df[[feat_i]].values,
                    mi_df[feat_j].values,
                    discrete_features=[feat_i in ["vtat_missing", "vehicle_type_enc", "pickup_enc", "drop_enc", "route_enc"]],
                    random_state=42, n_neighbors=5
                )[0]
            mi_matrix.iloc[i, j] = mi
            mi_matrix.iloc[j, i] = mi

print("Pairwise MI between features:")
print(mi_matrix.to_string())


Pairwise MI between features:
              avg_vtat  vtat_missing  vehicle_type    pickup      drop     route
avg_vtat           NaN      0.227924      0.000473  0.000000  0.000581  0.001518
vtat_missing  0.227924           NaN      0.000034  0.000598  0.000489  0.090601
vehicle_type  0.000473      0.000034           NaN  0.003507  0.003664  0.619317
pickup        0.000000      0.000598      0.003507       NaN  0.112739  5.169869
drop          0.000581      0.000489      0.003664  0.112739       NaN  5.169916
route         0.001518      0.090601      0.619317  5.169869  5.169916       NaN


Most feature pairs have near-zero MI, matching the VIF results, no hidden non-linear dependencies. Two exceptions stand out: pickup-drop and route-pickup-drop all show extremely high MI (5+), but this is a definitional artefact, route is deterministically constructed from pickup and drop so they share perfect mutual information by construction. The only genuine non-linear dependency is avg_vtat <> vtat_missing (MI = 0.23), which we already knew, vtat_missing flags where avg_vtat is NaN. vehicle_type <> route has moderate MI (0.62), meaning certain routes are more associated with specific vehicle types, but not enough to cause modeling issues. Everything else is independent

## Conclusions

**Final feature set going into modeling:**
- avg_vtat + vtat_missing (structurally complementary, non-redundant)
- vehicle_type
- pickup_location, drop_location
- route (candidate, needs validation)

**What this notebook established:**
- avg_vtat missingness is MNAR driven by the target with no hidden confounding
- vtat_zone is redundant with avg_vtat (eta-sq = 0.82), dropped
- is_instant_arrival, is_timeout, is_long_wait are deterministic subsets of avg_vtat, dropped
- Temporal families (hour, day_of_year and all derivatives) carry no meaningful signal, dropped entirely
- The pickup-drop route combination is a much stronger predictor than either location individually, but the Cramer's V is inflated by high cardinality (30k levels) and data sparsity (avg 5 rides per route). The signal exists but its robustness is unproven
- No multicollinearity among the selected features (all VIFs near 1)

**What remains for modeling:**
- Route needs cross-validated testing to confirm the signal survives out-of-sample
- Encoding strategy for route and locations (target encoding, frequency encoding, embeddings) is a modeling decision
- avg_vtat imputation strategy (median fill + vtat_missing flag vs sentinel value) depends on model family